src/github.rs

```diff
@@ -37,132 +69,129 @@ pub async fn auth() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
 
     let codes = crab
-        .authenticate_as_device(&secret_id, ["public_repo"])
+        .authenticate_as_device(&client_id, ["public_repo"])
         .await?;
 
     // Set message box text
     let message = format!(
-        "{}\n{}",
-        codes.verification_uri.clone(),
-        codes.user_code.clone()
+        "<a href=\"{0}\">{0}</a>\n{1}",
+        codes.verification_uri, codes.user_code
     );
 
-    let mut backend = dialog::backends::Zenity::new();
-    backend.set_width(250);
-    dialog::Message::new(message)
-        .title("Github Login")
-        .show_with(backend)
-        .expect("Could not display Github dialog box");
+    // rx.recv() will resolve when _tx is dropped
+    let (_tx, rx) = async_channel::bounded::<()>(1);
+    let (cancel_tx, cancel_rx) = async_channel::bounded::<()>(1);
+
+    // GObjects are not Send + Sync, hence cannot be held across await. First create a future that
+    // is run in main thread, and use the local variant from there.
+    gtk::glib::spawn_future(async move {
+        gtk::glib::spawn_future_local(async move {
+            let dlg = adw::MessageDialog::new(gtk::Window::NONE, Some("Github Login"), None);
+            let (cncl_tx, cncl_rx) = async_channel::bounded::<()>(1);
+            let _cancel_tx = cancel_tx;
+            dlg.set_body(&message);
+            dlg.add_response("cancel", "Cancel");
+            dlg.set_body_use_markup(true);
+            dlg.connect_response(None, move |_dlg, _ers| {
+                let _ = cncl_tx.send_blocking(());
+            });
+            dlg.show();
+            futures::select! {
+                _ = rx.recv().fuse() => (),
+                _ = cncl_rx.recv().fuse() => (),
+            };
+            dlg.destroy();
+        });
+    });
 
     // Atuhentication with timeout
-    let auth = tokio::time::timeout(timeout, codes.poll_until_available(&crab, &secret_id))
-        .await?
-        .unwrap();
-    // Write key to config file
-    set_key(auth.access_token.expose_secret()).unwrap();
+
+    let auth = tokio::select! {
+        e = codes.poll_until_available(&crab, &client_id) => e?,
+        _ = tokio::time::sleep(timeout) => Err(Error::TimedOut)?,
+        _ = cancel_rx.recv() => Err(Error::Cancelled)?,
+    };
+    set_key(auth.access_token)?;
+
     Ok(())
 }
 
-pub fn get_config_path() -> String {
+pub fn get_config_path() -> PathBuf {
     let variable_name = "GITHUB_CONFIG";
-    let variable = env::var(variable_name);
-    let path = match variable {
-        Ok(ref val) => val,
-        Err(e) => {
-            println!("Missing environment variable: {}, {}", variable_name, e);
-            "/home/ghaf/.config/ctrl-panel/config.toml"
+    let variable = env::var_os(variable_name);
+    match variable {
+        Some(ref val) => PathBuf::from(val),
+        None => {
+            println!("Missing environment variable: {variable_name}");
+            PathBuf::from("/home/ghaf/.config/ctrl-panel/config.toml")
```

Comment:
May be worth co build a path using `${HOME}/.config/....`
(non-blocking comment)